### PR TITLE
Refactor CVector/2D/4D to use constexpr + noexcept everywhere

### PR DESCRIPTION
**Pull request by Pirulax**
Basically this PR makes the above mentioned classes use constexpr, and noexcept everywhere they can.

All functions other than the constructors were probably inlined before anyways, but declaring the constructors as `constexpr` allows me(or someone else), to declare constexpr variables with vectors. 
Its kinda useful, just gives me that _constifying_ feel. 

These edits are cherry-picked from #1467

### CODE REVIEW

**Comment by Pirulax** [comment]
OMG IT WORKS ON THE FIRST TRY jeez m8 thats gr8

**Review by sbx320 — APPROVED** [review]

File: Shared/sdk/CVector.h [hunk 1]
```diff
@@ -26,53 +26,63 @@
...
     float Normalize()
     {
-        float t = sqrt(fX * fX + fY * fY + fZ * fZ);
+        const float t = Length();
         if (t > FLOAT_EPSILON)
         {
-            float fRcpt = 1 / t;
-            fX *= fRcpt;
-            fY *= fRcpt;
-            fZ *= fRcpt;
```
Comment (qaisjp) [inline Shared/sdk/CVector.h:60]: I thought muls were less expensive than divs, so this is an optimisation to div once, and mul multiple times.

It's possible that the compiler optimises this for us already, though.

File: Shared/sdk/CVector.h [hunk 2]
```diff
@@ -26,53 +26,63 @@
...
-            t = 0;
-        return t;
+            return 0;
     }
 
-    float Length() const { return sqrt((fX * fX) + (fY * fY) + (fZ * fZ)); }
+    inline float Length() const { return sqrt((fX * fX) + (fY * fY) + (fZ * fZ)); }
 
-    float LengthSquared() const { return (fX * fX) + (fY * fY) + (fZ * fZ); }
+    // returns just the squared length(eg.: x*x* + y*y + z*z)
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]:
```suggestion
    // LengthSquared returns Length() without sqrt applied (i.e. x*x* + y*y + z*z)
```

File: Shared/sdk/CVector.h [hunk 3]
```diff
@@ -26,53 +26,64 @@
...
-        return t;
+            return 0;
     }
 
-    float Length() const { return sqrt((fX * fX) + (fY * fY) + (fZ * fZ)); }
+    inline float Length() const { return sqrt((fX * fX) + (fY * fY) + (fZ * fZ)); }
 
-    float LengthSquared() const { return (fX * fX) + (fY * fY) + (fZ * fZ); }
+    // LengthSquared returns Length() without sqrt applied (i.e. returns x*x* + y*y + z*z)
+    // For more info see CMaterialLine3DBatcher::Flush()
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]: I don't think it's necessary to include the second line here because it's common to use `LengthSquared` for comparisons, which is when you don't strictly care about the square root.

If you still believe it's worth mentioning, then it might be clearer to have a comment like this, which is more descriptive and doesn't require the user to dig into potentially unrelated code.

```suggestion
    // LengthSquared returns Length() without sqrt applied (i.e. returns x*x* + y*y + z*z).
    // This can be useful if you only want to compare lengths.
```

File: Shared/sdk/CVector.h [hunk 4]
```diff
@@ -192,96 +195,88 @@ class CVector
...
-    CVector operator-() const { return CVector(-fX, -fY, -fZ); }
+    constexpr CVector operator*(const float fRight) const noexcept { return CVector(fX * fRight, fY * fRight, fZ * fRight); }
 
-    void operator+=(float fRight)
+
+    constexpr CVector operator/(const CVector& vecRight) const noexcept { return CVector(fX / vecRight.fX, fY / vecRight.fY, fZ / vecRight.fZ); }
+
+    constexpr CVector operator/(const float fRight) const noexcept { return CVector(fX / fRight, fY / fRight, fZ / fRight); }
+
+
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]:
```suggestion
```

File: Shared/sdk/CVector.h [hunk 5]
```diff
@@ -192,96 +195,88 @@ class CVector
...
-    void operator+=(const CVector& vecRight)
+    constexpr void operator+=(const CVector& vecRight) noexcept
     {
         fX += vecRight.fX;
         fY += vecRight.fY;
         fZ += vecRight.fZ;
     }
 
-    void operator-=(float fRight)
+
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]:
```suggestion
```

File: Shared/sdk/CVector.h [hunk 6]
```diff
@@ -192,96 +195,88 @@ class CVector
...
-    void operator-=(const CVector& vecRight)
+    constexpr void operator-=(const CVector& vecRight) noexcept
     {
         fX -= vecRight.fX;
         fY -= vecRight.fY;
         fZ -= vecRight.fZ;
     }
 
-    void operator*=(float fRight)
+
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]:
```suggestion
```

File: Shared/sdk/CVector.h [hunk 7]
```diff
@@ -192,96 +195,88 @@ class CVector
...
-    void operator*=(const CVector& vecRight)
+    constexpr void operator*=(const CVector& vecRight) noexcept
     {
         fX *= vecRight.fX;
         fY *= vecRight.fY;
         fZ *= vecRight.fZ;
     }
 
-    void operator/=(float fRight)
+
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]:
```suggestion
```

File: Shared/sdk/CVector4D.h [hunk 8]
```diff
@@ -71,100 +61,83 @@ class CVector4D
...
+    constexpr CVector4D operator-(const CVector4D& vecRight) const noexcept { return CVector4D(fX - vecRight.fX, fY - vecRight.fY, fZ - vecRight.fZ, fW - vecRight.fW); }
 
-    CVector4D operator/(const CVector4D& vecRight) const { return CVector4D(fX / vecRight.fX, fY / vecRight.fY, fZ / vecRight.fZ, fW / vecRight.fW); }
 
-    void operator+=(float fRight)
+    constexpr CVector4D operator/(const CVector4D& vecRight) const noexcept { return CVector4D(fX / vecRight.fX, fY / vecRight.fY, fZ / vecRight.fZ, fW / vecRight.fW); }
+
+    constexpr CVector4D operator/(const float fRight) const noexcept { return CVector4D(fX / fRight, fY / fRight, fZ / fRight, fW / fRight); }
+
+
```
Comment (qaisjp) [inline Shared/sdk/CVector4D.h]:
```suggestion
```

**Review by qaisjp — CHANGES_REQUESTED** [review]
Thank you! I've left a couple style comments, a comment comment, and a suggestion to keep one potential optimisation.

File: Shared/sdk/CVector.h [hunk 9]
```diff
@@ -26,53 +26,64 @@
...
+        fX = vec.fX;
+        fY = vec.fY;
+        fZ = vec.fZ;
+        return *this;
     }
 
+
+    constexpr CVector Clone() const { return *this; }
+
+    // returns the length of the vector and normalizes it
```
Comment (qaisjp) [inline Shared/sdk/CVector.h]:
```suggestion
    // Normalize returns the normalized length of the vector.
```

**Comment by Pirulax** [comment]
I feel like that "potential optimization" is quite unnecessary, as, by default MSVC uses SSE2, which will probably dump all the 4 floats into a m128 reg, and divide them. Which wont be much slower than just doing this optimization trick(which we shoudnt, as the compiler will optimize it, especially if we enable fast-math(which will pretty much guarantee us that the compiler will use SSE2 with this trick we've used:
``` cpp
#pragma fast-math push
#pragma fast-math on
// code
#pragma fast-math pop
```

File: Shared/sdk/CVector2D.h [hunk 10]
```diff
@@ -20,128 +20,123 @@
...
 
-    float DotProduct(CVector2D& other) const { return fX * other.fX + fY * other.fY; }
+    constexpr float DotProduct(CVector2D& other) const { return fX * other.fX + fY * other.fY; }
 
-    float Length() const { return sqrt(fX * fX + fY * fY); }
+    float Length() const { return std::hypotf(fX, fY); }
 
-    float LengthSquared() const { return (fX * fX) + (fY * fY); }
+    // LengthSquared returns Length() without sqrt applied (i.e. returns x*x* + y*y)
+    // For more info see CMaterialLine3DBatcher::Flush()
```
Comment (qaisjp) [inline Shared/sdk/CVector2D.h]:
```suggestion
    // LengthSquared returns Length() without sqrt applied (i.e. returns x*x* + y*y).
    // This can be useful if you only want to compare lengths.
```

File: Shared/sdk/CVector4D.h [hunk 11]
```diff
@@ -20,48 +20,38 @@
...
+    // Warning, this function is returning the wrong value(fW is missing), its kept because nothing uses it, only
+    // CLuaVector4DDefs.
+    constexpr float DotProduct(const CVector4D& other) const noexcept { return fX * other.fX + fY * other.fY + fZ * other.fZ; }
 
-    float Length() const { return sqrt(fX * fX + fY * fY + fZ * fZ + fW * fW); }
+    float Length() const noexcept { return sqrt(fX * fX + fY * fY + fZ * fZ + fW * fW); }
 
-    float LengthSquared() const { return (fX * fX) + (fY * fY) + (fZ * fZ) + (fW * fW); }
+    // LengthSquared returns Length() without sqrt applied (i.e. returns x*x* + y*y + z*z + w*w)
+    // For more info see CMaterialLine3DBatcher::Flush()
```
Comment (qaisjp) [inline Shared/sdk/CVector4D.h]:
```suggestion
    // LengthSquared returns Length() without sqrt applied (i.e. returns x*x* + y*y + z*z + w*w).
    // This can be useful if you only want to compare lengths.
```

**Review by qaisjp — APPROVED** [review]

**Comment by ghost** [comment]
@Pirulax 
> tbh I agree, but nobody reads CVector code anyways. Constexpr is needed(at least for the constructor) in CGraphics refactor(somewhere in the end, the sphere thing)

Why is constexpr needed there? The main purpose of constexpr is to compute the result in compile time. That can only happen if the compiler knows about the values used in the method at compile-time, otherwise, it will just inline it at best and won't do any good optimizations.

**Comment by Pirulax** [comment]
As stated somewhere, I need this for the CGraphics refactor thingy
